src/ripple/overlay/impl/PeerImp.cpp

```diff
@@ -1965,6 +2009,134 @@ PeerImp::onMessage (std::shared_ptr <protocol::TMHaveTransactionSet> const& m)
+            remote_address_.to_string() << " (" << id_ << ") with result " <<
+            to_string(disp);
+
+        switch (disp)
+        {
+        case ListDisposition::accepted:
+            JLOG (p_journal_.debug()) <<
+                "Applied new validator list from peer " << remote_address_;
+            {
+                // if the remote node uses any of the same lists
+                // as this node, treat it as sane.
+                auto expected = Sanity::unknown;
+                sanity_.compare_exchange_strong(expected, Sanity::sane);
+
+                std::lock_guard<std::mutex> sl(recentLock_);
+
+                assert(applyResult.sequence && applyResult.publisherKey);
+                auto const& pubKey = *applyResult.publisherKey;
+#ifndef NDEBUG
+                if (auto const iter = publisherListSequences_.find(pubKey);
+                    iter != publisherListSequences_.end())
+                {
+                    assert(iter->second < *applyResult.sequence);
+                }
+#endif
+                publisherListSequences_[pubKey] = *applyResult.sequence;
+            }
+            break;
+        case ListDisposition::same_sequence:
+            JLOG (p_journal_.warn()) <<
+                "Validator list with current sequence from peer " <<
+                remote_address_;
+            // Charging this fee here won't hurt the peer in the normal
+            // course of operation (ie. refresh every 5 minutes), but
+            // will add up if the peer is misbehaving.
+            fee_ = Resource::feeUnwantedData;
+#ifndef NDEBUG
+            {
+                std::lock_guard<std::mutex> sl(recentLock_);
+                assert(applyResult.sequence && applyResult.publisherKey);
+                assert(publisherListSequences_[*applyResult.publisherKey]
+                    == *applyResult.sequence);
+            }
+#endif // !NDEBUG
+
+            break;
+        case ListDisposition::stale:
+            JLOG (p_journal_.warn()) <<
+                "Stale validator list from peer " << remote_address_;
+            // There are very few good reasons for a peer to send an
+            // old list, particularly more than once.
+            fee_ = Resource::feeBadData;
+            break;
+        case ListDisposition::untrusted:
+            JLOG (p_journal_.warn()) <<
+                "Untrusted validator list from peer " << remote_address_;
+            // Charging this fee here won't hurt the peer in the normal
+            // course of operation (ie. refresh every 5 minutes), but
+            // will add up if the peer is misbehaving.
+            fee_ = Resource::feeUnwantedData;
+            break;
+        case ListDisposition::invalid:
+            JLOG (p_journal_.warn()) <<
+                "Invalid validator list from peer " << remote_address_;
+            // This shouldn't ever happen with a well-behaved peer
+            fee_ = Resource::feeInvalidSignature;
```

Comment:
I think that this warrants disconnecting the peer on the spot. Rationale: the peer has _already_ sent us invalid data; what's the point of giving it a chance to send us more invalid data?

Reply:
By that logic, shouldn't we drop the connection (almost?) everywhere we charge `feeInvalidSignature`? If that's the case, why not change that cost from 1000 to 3000?

Reply:
You're right that in some of those cases, dropping would be appropriate. It's nothing urgent, we can leave it as is now, and then decide how we want to tackle this (i.e. charge vs. drop) across the board in a subsequent PR.